lib/Dialect/ModArith/IR/ModArithDialect.cpp

```diff
@@ -128,6 +136,50 @@ LogicalResult BarrettReduceOp::verify() {
   return success();
 }
 
+ParseResult ConstantOp::parse(OpAsmParser &parser, OperationState &result) {
+  APInt parsedValue(64, 0);
+  Type parsedType;
+
+  if (failed(parser.parseInteger(parsedValue))) {
+    parser.emitError(parser.getCurrentLocation(),
+                     "found invalid integer value");
+    return failure();
+  }
+
+  if (parser.parseColon() || parser.parseType(parsedType)) return failure();
+
+  auto modArithType = dyn_cast<ModArithType>(parsedType);
+  if (!modArithType) return failure();
+
+  auto outputBitWidth =
+      modArithType.getModulus().getType().getIntOrFloatBitWidth();
+  if (parsedValue.getActiveBits() > outputBitWidth)
+    return parser.emitError(parser.getCurrentLocation(),
+                            "constant value is too large for the modulus");
+
+  auto intValue = IntegerAttr::get(modArithType.getModulus().getType(),
+                                   parsedValue.trunc(outputBitWidth));
+  result.addAttribute(
+      "value", ModArithAttr::get(parser.getContext(), modArithType, intValue));
+  result.addTypes(modArithType);
+  return success();
+}
+
+void ConstantOp::print(OpAsmPrinter &p) {
+  p << " ";
+  getValue().getValue().getValue().print(p.getStream(), true);
```

Comment:
Looks like vscode swallowed the one comment actually related to this PR:
`getValue().getValue().getValue()` is pretty cursed xD  Maybe add a comment to indicate what each level is actually getting (Op's "value" attribute, the APInt in the attribute, the integer in in the APInt I'm guessing?)